### DISCUSSION TITLE
How to solve the 2-D Navier-Stokes equation with py-pde?

### DISCUSSION THREAD

**Discussion by KEUNGHE** [zwicker-group/py-pde#502] [Q&A]
I want to solve the 2-d Navier-Stokes equation for a viscous, incompressible fluid in vorticity form. But I don't know how to implement it with py-pde since I didn't find a similar example. Is it possible to solve this equaiton?
![Equation](https://github.com/zwicker-group/py-pde/assets/96622224/0b05c0d6-36ff-4cf5-9841-3cf3ed31da27)

**Comment by david-zwicker** [comment]
I haven't thought about it, but it might be possible.

**Reply by KEUNGHE** [reply]
Thank you for your reply. I'm going to give it a try.

**Comment by sijieh23** [comment]
I'm looking for this possibility as well. Though I'm more interested in its potential to solve 3D NS equation.

I just stumbled upon this code, so haven't yet look deep into it. On the face of it, I have following concerns/thoughts/...
1. I don't see linear solver implemented, nor multigrid-type of method, for solving pressure Poisson equation 
2. Don't know if it's easy to implement staggered-grid configuration, which is essential for incompressible flow simulation 
3. It'd be great if there are some results of scalability tests (strong/weak), to get some idea of the possibility running large simulations using the code

**Reply by david-zwicker** [reply]
You are right that those features are not currently supported. It is unlikely that we will implement this in the near future since it would be a substantial developmental burden. However, you could try to implement your own pressure solver (e.g. using spectral methods) and then use the rest of the infrastructure of `py-pde` to advance the system in time. I'd be happy to look at a pull request that implements such a PDE!

**Reply by sijieh23** [reply]
> However, you could try to implement your own pressure solver (e.g. using spectral methods) and then use the rest of the infrastructure of py-pde to advance the system in time

Interesting! That's something I should think about.

Though I have a few more questions:

1. What's your domain decomposition strategy? How you exchange data for the ghost points? I really appreciate it if you can point me the code so that I can take a look.
2. Did you perform scalability tests? Is it scalable to large-size simulation?

**Reply by david-zwicker** [reply]
Since all grids are orthogonal, we simply decompose them in blocks (chosen by the user). I guess one could add some automatic decomposition strategy, but we haven't implemented this. The ghost points are exchanged at every time step using special boundary conditions. In essence, when a ghost point is required in the simulation, the neighboring node is asked to provide it (this is vectorized for speed). In the end all nodes then exchange ghost points synchronously. This strategy won't directly work when you want to solve a Poisson equation (like the pressure used above).
